### DISCUSSION TITLE
--watch with .ts file?

### DISCUSSION THREAD

**Discussion by nshiab** [nodejs/help#4593] [Q&A] [answered]
I often use .ts files with bun --watch, because it's handy. But I wonder if there is an easy way to do the same thing with nodejs.

So I am curious: how do you usually watch .ts files and run them?

Thank you! 🙏

**Comment by avivkeller** [comment]
Node.js doesn't provide *native* support for TypeScript, but there may be loaders that do, if I'm not mistaken.

The project is discussing how to incorperate TypeScript into Node.js at https://github.com/nodejs/node/issues/43816 and https://github.com/nodejs/node/issues/43818

**Reply by avivkeller** [reply]
Marked as answer, but LMK if you disagree and I'll change it

**Reply by bmuenzenmeyer** [reply]
since i found this discussion myself via Google, it'd be nice to edit this answer or change it to mine at https://github.com/orgs/nodejs/discussions/53624#discussioncomment-11717154

**Comment by nshiab** [comment]
Looks like you can:
- Install ts-node
- Install node-dev
- And then run `npx node-dev --respawn index.ts`

More steps than Bun or Deno, but still fairly easy...

**Reply by avivkeller** [reply]
Thanks for the insight! I've marked the other response as the answer, because this question specifically refers to --watch, but this suggestion is also very good :-)

**Reply by shinebayar-g** [reply]
ts-node often breaks when CommonJS/ESM packages are involved. At some point, everyone will face the famous `ERR_UNKNOWN_FILE_EXTENSION` error and can't figure out why.

It's just too complicated to set up if you're not a Node.js hacker. tsx is somewhat better but still a hacky solution. I wish there was native Node.js support for TypeScript.

To be honest, frontend developers who are shipping JavaScript to browsers have already figured out how to deal with it and have great bundlers. But when Node.js is solely being used in the backend, for scripting, or in an infrastructure context, we don't care about browsers and bundlers for JavaScript. We just want Typescript and NodeJS to work.

**Reply by avivkeller** [reply]
The project is discussing how to incorperate TypeScript into Node.js at #43816 and #43818

(Edit: I see you are heavily involved in that discussion, great work, and I look forward to seeing two projects come together)

**Reply by shinebayar-g** [reply]
Yeah. I refresh those issues almost daily to see if something new comes up. :pray:

**Reply by avivkeller** [reply]
Best of luck :heart:

**Reply by nshiab** [reply]
Ah! I subscribed to these two issues. I hope it will happen.

**Comment by bmuenzenmeyer** [comment, marked as answer]
The new answer, in 2025 and beyond, is that this works:

```
node --experimental-strip-types --watch-path src src/server.ts
```
